src/components/tl/cuda/alltoall/alltoall_ce.c

```diff
@@ -0,0 +1,250 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "alltoall.h"
+#include "core/ucc_mc.h"
+#include "tl_cuda_cache.h"
+#include "tl_cuda_topo.h"
+#include "utils/arch/cpu.h"
+
+enum {
+    ALLTOALL_CE_STAGE_SYNC, /*< Wait for free SYNC segment */
+    ALLTOALL_CE_STAGE_COPY, /*< Wait for all copies to finish */
+    ALLTOALL_CE_STAGE_BAR,  /*< Wait for other ranks to finish */
+};
+
+ucc_status_t ucc_tl_cuda_alltoall_ce_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_mc_ee_destroy_event((void*)task->alltoall_ce.copy_done,
+                            UCC_EE_CUDA_STREAM);
+    ucc_tl_cuda_task_put(task);
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_cuda_alltoall_setup(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t          *team = TASK_TEAM(task);
+    ucc_tl_cuda_sync_t          *sync = TASK_SYNC(task, UCC_TL_TEAM_RANK(team));
+    volatile ucc_tl_cuda_sync_t *peer_sync;
+    ucc_tl_cuda_cache_t         *cache;
+    ucc_status_t                 status;
+    ucc_rank_t                   i;
+
+    memcpy(&sync->mem_info_src, &task->alltoall_ce.mem_info_src,
+           sizeof(ucc_tl_cuda_mem_info_t));
+    memcpy(&sync->mem_info_dst, &task->alltoall_ce.mem_info_dst,
+           sizeof(ucc_tl_cuda_mem_info_t));
+
+    CUDACHECK_GOTO(cudaEventRecord(sync->ipc_event_local, team->stream),
+                   exit_err, status, UCC_TL_TEAM_LIB(team));
+    ucc_memory_bus_fence();
+    status = ucc_tl_cuda_shm_barrier_start(UCC_TL_TEAM_RANK(team), task->bar);
+    if (ucc_unlikely(status != UCC_OK)) {
+        goto exit_err;
+    }
+    while (ucc_tl_cuda_shm_barrier_test(UCC_TL_TEAM_RANK(team), task->bar) ==
+           UCC_INPROGRESS);
+    if (ucc_unlikely(status != UCC_OK)) {
+        goto exit_err;
+    }
+
+    for (i = 0; i < UCC_TL_TEAM_SIZE(team); i++) {
+        if (i == UCC_TL_TEAM_RANK(team) ||
+            !ucc_tl_cuda_topo_is_direct(team, team->topo, UCC_TL_TEAM_RANK(team), i)) {
+            continue;
+        }
+        peer_sync = TASK_SYNC(task, i);
+        cache = ucc_tl_cuda_get_cache(team, i);
+        if (ucc_unlikely(!cache)) {
+            status = UCC_ERR_NO_MESSAGE;
+            goto exit_err;
+        }
+        status = ucc_tl_cuda_map_memhandle(peer_sync->mem_info_src.ptr,
+                                           peer_sync->mem_info_src.length,
+                                           peer_sync->mem_info_src.handle,
+                                           &task->alltoall_ce.peer_map_addr_src[i],
+                                           cache);
+        if (UCC_OK != status) {
+            ucc_error("ucc_cuda_ipc_map_memhandle failed");
+            return UCC_ERR_INVALID_PARAM;
+        }
+        status = ucc_tl_cuda_map_memhandle(peer_sync->mem_info_dst.ptr,
```

Comment:
so now you have to map both src and dst? is this needed only for proxy or always? does this imply overheads when "proxy" is not required ?

Reply:
right, dst buffer registration is needed only if we use proxy rank however with registration cache overhead is not big. Let's postpone this change for another PR, there are some other optimization I need to add